server/proxy/bindings.go

```diff
@@ -86,122 +91,156 @@ func (p *Proxy) GetAppBindings(in Incoming, cc apps.Context, app apps.App) []app
 
 	case apps.CallResponseTypeError:
-		log.WithError(resp).Debugf("Error getting bindings")
-		return nil
+		problems = append(problems, errors.Wrap(resp, "received app error"))
+		return nil, problems
 
 	default:
-		log.Debugf("Bindings response is nil or unexpected type.")
-		return nil
+		problems = append(problems, errors.Errorf("unexpected response type %q", string(resp.Type)))
+		return nil, problems
 	}
 }
 
-// scanAppBindings removes bindings to locations that have not been granted to
+// cleanAppBindings removes bindings to locations that have not been granted to
 // the App, and sets the AppID on the relevant elements.
-func (p *Proxy) scanAppBindings(app apps.App, bindings []apps.Binding, locPrefix apps.Location, userAgent string) []apps.Binding {
+func cleanAppBindings(app apps.App, bindings []apps.Binding, locPrefix apps.Location, userAgent string, conf config.Config) ([]apps.Binding, []error) {
 	out := []apps.Binding{}
-	locationsUsed := map[apps.Location]bool{}
-	labelsUsed := map[string]bool{}
-	conf, _, log := p.conf.Basic()
-	log = log.With("app_id", app.AppID)
+	usedLocations := map[apps.Location]bool{}
+	usedCommandLabels := map[string]bool{}
 
+	var problems []error
 	for _, b := range bindings {
-		if b.Location == "" {
-			b.Location = apps.Location(app.Manifest.AppID)
+		clean, appProblems := cleanAppBinding(app, b, locPrefix, userAgent, conf)
+		problems = append(problems, appProblems...)
+		if clean == nil {
+			continue
 		}
 
-		fql := locPrefix.Make(b.Location)
-		allowed := false
-		for _, grantedLoc := range app.GrantedLocations {
-			if fql.In(grantedLoc) || grantedLoc.In(fql) {
-				allowed = true
-				break
-			}
+		fql := locPrefix.Sub(clean.Location)
+		if usedLocations[clean.Location] {
+			problems = append(problems,
+				errors.Errorf("ignored diplicate command binding for location %q", clean.Location))
+			continue
 		}
-		if !allowed {
-			log.Debugw("location is not granted to app", "location", fql)
+		if fql.In(apps.LocationCommand) && usedCommandLabels[clean.Label] {
+			problems = append(problems,
+				errors.Errorf("ignored diplicate command binding for label %q (location %q)", clean.Label, clean.Location))
 			continue
 		}
 
-		if fql.In(apps.LocationCommand) {
-			label := b.Label
-			if label == "" {
-				label = string(b.Location)
-			}
+		out = append(out, *clean)
+	}
 
-			if strings.ContainsAny(label, " \t") {
-				log.Debugw("Binding validation error: Command label has multiple words", "location", b.Location)
-				continue
-			}
-		}
+	return out, problems
+}
 
-		if fql.IsTop() {
-			if locationsUsed[b.Location] {
-				continue
-			}
-			locationsUsed[b.Location] = true
-		} else {
-			if b.Location == "" || b.Label == "" {
-				continue
-			}
-			if locationsUsed[b.Location] || labelsUsed[b.Label] {
-				continue
-			}
+func cleanAppBinding(
+	app apps.App,
+	b apps.Binding,
+	locPrefix apps.Location,
+	userAgent string,
+	conf config.Config,
+) (*apps.Binding, []error) {
```

Comment:
Should we use https://pkg.go.dev/github.com/hashicorp/go-multierror here as it stick with the standard of returning `error` and it's already used in `apps`?